### DISCUSSION TITLE
How to define capsules declared in another package?

### DISCUSSION THREAD

**Discussion by busslina** [GregoryConrad/rearch-dart#85] [Q&A] [answered]
This is the scenario: I have a package library and a project. In the package library I define an abstract (or mixin) `App` entity that must be implemented/extended on the project. Also I want to be able to access the `App` instance capsule all over the package library.

Can you give me some solutions to this?

**Comment by busslina** [comment]
I did this solution, but maybe there is a better/simpler solution:

```dart
AppController appControllerCapsule(CapsuleHandle use) {
  final (getInitialized, setInitialized) = use.rawValueWrapper(() => false);
  final (getApp, setApp) = use.rawValueWrapper<WebApp>();

  return (
    get: getApp,
    set: (app) {
      if (getInitialized()) {
        throw ('appControllerCapsule -- Yet initialized');
      }

      setApp(app);

      setInitialized(true);
    },
  );
}

WebApp get app => cc.read(appControllerCapsule).get();

abstract class WebApp {

  WebApp() {
    // Saving instance
    cc.read(appControllerCapsule).set(this);
  }
}
```

**Comment by GregoryConrad** [comment, marked as answer]
> In the package library I define an abstract (or mixin) App entity that must be implemented/extended on the project.

That goes against ReArch quite a bit (using inheritance). You should instead expose capsules in your library package that declare a certain API to be defined elsewhere so that you can then use those common capsules in your library package.

There are two ways I can think of to do this off the top of my head.
1. (Not preferred) Use a singleton/static instance that allows you to set the current implementation of some capsule.
2. (Preferred) Make your library package's capsules act as "proxies" that route all requests to another capsule. Here's an example:

```dart
// In your library package:
(Capsule<int>, void Function(Capsule<int>)) _countCapsuleManager(CapsuleHandle use) =>
    use.state((CapsuleHandle use) => throw UnsupportedError('Must set the countCapsule via countCapsuleSetterAction!'));
void Function(Capsule<int>) countCapsuleSetterAction(CapsuleHandle use) => use(_countCapsuleManager).$2;
int countCapsule(CapsuleHandle use) => use(use(_countCapsuleManager).$1); // Your proxy

// Somewhere shortly after you make your container in your application's package:
container.read(countCapsuleSetterAction)(theActualCountCapsuleImpl);
```

**Reply by GregoryConrad** [reply]
Worth noting that you _must_ set the `countCapsule` from above with `countCapsuleSetterAction` before ever `use`-ing the `countCapsule` (even indirectly) in your application code. Thus, the setting of the capsule should be the first thing you do after creating the container.

**Reply by busslina** [reply]
Cool solution. Thank you !!
